File: src/jsonrpc/eth/uninstallFilter/eth_uninstallFilter.go
package eth

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Uninstalls a filter with given id.
//
// Example:
// Filter identifier: "0x01"
// Result: true
//
// Implements the eth_uninstallFilter JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "eth_uninstallFilter"

// Params represents the parameters for eth_uninstallFilter
type Params struct {
	// hex encoded unsigned integer
	Filter identifier types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
		p.Filter identifier,
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 1 {
		return fmt.Errorf("expected 1 parameters, got %d", len(arr))
	}

	if err := json.Unmarshal(arr[0], &p.Filter identifier); err != nil {
		return fmt.Errorf("parameter 0 (Filter identifier): %w", err)
	}

	return nil
}

// Result represents the result for eth_uninstallFilter
type Result struct {
	Value types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}

File: src/jsonrpc/eth/uninstallFilter/eth_uninstallFilter.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Uninstalls a filter with given id.
///
/// Example:
/// Filter identifier: "0x01"
/// Result: true
///
/// Implements the `eth_uninstallFilter` JSON-RPC method.
pub const EthUninstallFilter = @This();

/// The JSON-RPC method name
pub const method = "eth_uninstallFilter";

/// Parameters for `eth_uninstallFilter`
pub const Params = struct {
    /// hex encoded unsigned integer
    filter_identifier: types.Quantity,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.filter_identifier);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 1) return error.InvalidParamCount;

        return Params{
            .filter_identifier = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[0], options),
        };
    }
};

/// Result for `eth_uninstallFilter`
pub const Result = struct {
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};
